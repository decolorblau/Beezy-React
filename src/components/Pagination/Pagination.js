import { useContext, useState } from "react";
import "./Pagination.scss";
import MarvelContext from "../../store/contexts/MarvelContext";

const Pagination = () => {
  const {
    limit,
    offsetCharacters,
    setOffsetCharacters,
    offsetEvents,
    setOffsetEvents,
    offsetComics,
    setOffsetComics,
    title,
  } = useContext(MarvelContext);
  const { counter, setCounter } = useState(0);
  const { total, setTotal } = useState(74);

  const beforePage = () => {
    switch (title) {
      case "EVENTS":
        if (offsetEvents > 0) {
          setOffsetEvents(offsetEvents - limit);
          setCounter(offsetEvents);
          setTotal(74);
        }
        break;
      case "CHARACTERS":
        if (offsetCharacters > 0) {
          setOffsetCharacters(offsetCharacters - limit);
          setCounter(offsetCharacters);
          setTotal(1559);
        }
        break;
      case "COMICS":
        if (offsetComics > 0) {
          setOffsetComics(offsetComics - limit);
          setCounter(offsetComics);
          setTotal(51242);
        }
        break;
      default:
        return "error";
    }
  };

  const nextPage = () => {
    switch (title) {
      case "EVENTS":
        if (offsetEvents < 74) {
          setOffsetEvents(offsetEvents + limit);
          setCounter(offsetEvents);

          setTotal(74);
        }
        break;
      case "CHARACTERS":
        if (offsetCharacters < 1559) {
          setOffsetCharacters(offsetCharacters + limit);
          setCounter(offsetCharacters);
          setTotal(1559);
        }
        break;
      case "COMICS":
        if (offsetComics < 51242) {
          setOffsetComics(offsetComics + limit);
          setTotal(51242);
        }
        break;
      default:
        return "error";
    }
  };

  return (
    <>
      <div className="pagination">
        <button onClick={beforePage}>{"❮❮"} Before</button>
        <p>
          {counter} - {counter + limit} of {total}
        </p>
        <button onClick={nextPage}>Next {">>"}</button>
      </div>
    </>
  );
};

export default Pagination;
