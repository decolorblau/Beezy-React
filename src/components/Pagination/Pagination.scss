.pagination {
  display: flex;
  justify-content: space-between;
  width: 90%;
  max-width: 1440px;
  padding: 20px 0;
  color: white;
  align-items: center;
  button {
    font-family: "Roboto", sans-serif;
    font-size: 16px;
    color: #dcdcdc;
    float: left;
    padding: 8px 16px;
    text-decoration: none;
    height: 30px;
    background-color: transparent;
    border: none;
  }
  p {
    font-size: 14px;
  }
}

button:hover {
  color: rgb(209, 0, 0);
}

button:disabled {
  color: #585858;
}
