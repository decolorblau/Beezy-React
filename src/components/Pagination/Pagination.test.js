import Pagination from "./Pagination";
import { fireEvent, render, screen, cleanup } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import MarvelContext from "../../store/contexts/MarvelContext";
import ReactTestRenderer from "react-test-renderer";

beforeEach(cleanup);

describe("Given the component Pagination", () => {
  describe("When it receives an object with limit, offset, setOffset and total", () => {
    test("Then it should render pagination component", () => {
      const paginationValues = {
        limit: 20,
        offset: 0,
        setOffset: 20,
        total: 40,
      };

      const paginationComponent = ReactTestRenderer.create(
        <MarvelContext.Provider value={paginationValues}>
          <Router>
            <Pagination />
          </Router>
        </MarvelContext.Provider>
      );

      expect(paginationComponent.toJSON()).toMatchSnapshot();
    });
    test("Then it should be a button with comic text, another with magazine text an another with trade paperback", () => {
      const paginationValues = {
        limit: 20,
        offset: 0,
        setOffset: 20,
        total: 40,
      };
      const expected = "0 - 20 of 40 comics";

      render(
        <MarvelContext.Provider value={paginationValues}>
          <Router>
            <Pagination />
          </Router>
        </MarvelContext.Provider>
      );

      const infoComics = screen.getByTitle("info-comics").textContent;

      expect(infoComics).toContain(expected);
    });
    describe("And When Next button is clicked", () => {
      test("Then it should call nextPage function", () => {
        const paginationValues = {
          limit: 20,
          offset: 0,
          setOffset: jest.fn(),
          total: 40,
        };
        render(
          <MarvelContext.Provider value={paginationValues}>
            <Router>
              <Pagination />
            </Router>
          </MarvelContext.Provider>
        );

        const nextButton = screen.getByRole("button", {
          name: "Next >>",
        });
        expect(nextButton).toBeInTheDocument();

        fireEvent.click(nextButton);

        expect(paginationValues.setOffset).toHaveBeenCalled();
      });
    });
    describe("And When Before button is clicked", () => {
      test("Then it should call nextPage function", () => {
        const paginationValues = {
          limit: 20,
          offset: 20,
          setOffset: jest.fn(),
          total: 40,
        };
        render(
          <MarvelContext.Provider value={paginationValues}>
            <Router>
              <Pagination />
            </Router>
          </MarvelContext.Provider>
        );

        const beforeButton = screen.getByRole("button", {
          name: "<< Before",
        });

        expect(beforeButton).toBeInTheDocument();

        fireEvent.click(beforeButton);

        expect(paginationValues.setOffset).toHaveBeenCalled();
      });
    });
  });
});
