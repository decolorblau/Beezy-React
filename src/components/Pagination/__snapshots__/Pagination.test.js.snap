// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the component Pagination When it receives an object with limit, offset, setOffset and total Then it should render pagination component 1`] = `
<div
  className="pagination"
>
  <button
    disabled={true}
    onClick={[Function]}
  >
    &lt;&lt;
     Before
  </button>
  <p
    title="info-comics"
  >
    0
     - 
    20
     of 
    40
     comics
  </p>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Next 
    &gt;&gt;
  </button>
</div>
`;
